Document custom-kubernetes usage (#733)
akka/akka-platform-guide: docs-source/docs/modules/deployment/pages/bring-your-kubernetes-cluster.adoc

= Custom Kubernetes Deployment

include::partial$include.adoc[]

Users with a

> {lightbend-developer-docs}/introduction/getting-started/subscription-and-credentials.html[Lightbend Subscription {tab-icon}, window="tab"] have access to the Akka Cloud Platform operator to run on their own Kubernetes cluster.

Reviewer: We should document our versions policy for Kubernetes. The doc had some reference to version-skew-policy, likely we want to be even more restrictive than the Kubernetes policy: https://docs.google.com/document/d/11kDVXyi3qn3xnpUI3Az1zm-VfPPS9iriOp3IkreyjC4/edit#

The doc is currently out of date, but I think this part might still valid, let's check with @jtownley .

Reviewer 2: I agree we should mention this. @Ray Why more restrictive? What are you thinking?

Author: I'm not familiar enough with Kubernetes versions and the requirements for the operator to work as expected. I think we could specify that Kubernetes 1.19 is required and be a bit vague so the write-up of supported Kubernetes versions is easier to maintain.

Eventually, when we're swamped in support cases we'll have a wider testbed and can better document the supported versions or the minimum k8s version for each operator release.

Author: In any case, we don't specify a Kubernetes version anywhere else on this guide. IIRC, GKE supports specifying the k8s version to use when creating a cluster and we don't set a minimum version.

Currently the only reference to a k8s version is on a link to some docs (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#event-v1-core) but those docs are identical in 1.18.

We should have a compatibility matrix for the operator/k8s/cloud/metrics-server combinations but I'd rather do it on a separate PR.

Reviewer 3: I guess that we want to include support for Kubernetes versions that are not yet supported by GKE and EKS, but do we want to drop support for versions that they still support? 

I'm asking because following the Kubernetes version-skew-policy may not be the best option, considering how cloud providers pile up their own policies on top of that. For example, this is what [GKE docs says about version support](https://cloud.google.com/kubernetes-engine/versioning#stable):

> Google provides a total of **14 months of support** for each GKE minor version once the version has been made available in the **Regular** channel.

And here is the release schedule:

https://cloud.google.com/kubernetes-engine/docs/release-schedule

What we have there for K8s 1.19 is:

| Kubernetes version | Kubernetes release date | Rapid      | Regular    | Stable  |
|:-------------------|:------------------------|:-----------|:-----------|:--------|
| 1.19               | 2020-08-26              | 2021-01-25 | 2021-04-18 | 2021-06 |

It took ~5 months from K8s release to **Rapid** channel, then ~3 months from **Rapid** to **Regular**. From **Regular** to EOL, it will take 14 months. In total, from OSS release to EOL in GKE, it will take ~21 months.

Numbers for EKS are similar: https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html#kubernetes-release-calendar.

With a similar policy (from the FAQ in the link above):

> Q: How long is a Kubernetes version supported by Amazon EKS?
>
> A: A Kubernetes version is fully supported for **14 months after first being available on Amazon EKS**. This is true even if upstream Kubernetes is no longer supporting a version available on Amazon EKS.

I'm not sure what would be the best option, here, but a suggestion would be to support the union of:

1. K8s OSS supported versions
2. GKE versions in the *Regular* channel
3. Some subset of EKS supported versions?

WDYT?

Author: I think we can move this to a separate PR. As you surfaced, the issue is tricky and requires a bit more writeup and explanation than a simple paragraph.

Author: Raised #737

Make sure you read the xref:cloud-deployment.adoc[] and xref:deploy.adoc[] before proceeding to understand the basic deployment architecture and workflow.

== Prepare your cluster

Make sure your `kubectl` context points to your local Kubernetes cluster (for example, {minikube}[minikube {tab-icon}, window="tab"]), or an external Kubernetes cluster (e.g. in https://docs.microsoft.com/en-ca/azure/aks/kubernetes-walkthrough[Azure {tab-icon}, window="tab"]).

> Next, make sure you have the commercial Lightbend charts:

Reviewer: ```suggestion
Next, make sure you have the akka-operator-helm chart:
```
The charts are not really 'commercial'

[source,shell script]
----
helm repo add akka-operator-helm https://lightbend.github.io/akka-operator-helm/
helm repo update
----

Create environment variables with your {lightbend-com-account}/lightbend-platform/loginCredentials[Lightbend Login Credentials {tab-icon}, window="tab"]:

[source,shell script]
----
export LIGHTBEND_COMMERCIAL_USERNAME=12345678-aaa-bbb-cccc-ddddeeeeff0e@lightbend
export LIGHTBEND_COMMERCIAL_PASSWORD=abcdef1243568790ababab470556835429873569
----

Install the operator with the command:

[source,shell script,subs="attributes+"]
----
helm upgrade akka-operator akka-operator-helm/akka-operator \
  --set lightbendSubscription.username=$LIGHTBEND_COMMERCIAL_USERNAME \
  --set lightbendSubscription.password=$LIGHTBEND_COMMERCIAL_PASSWORD \
  --namespace lightbend \
  --version={akka-cloud-platform-version}  # <1>
----
<1> `{akka-cloud-platform-version}` is the latest available version.

The cluster is now ready for you to xref:deploy.adoc[deploy] `AkkaMicroservice` services.

== Known limitations

> The Akka Cloud Platform operator will not be in charge of creating the necessary `Ingress` instances to handle external traffic into your `AkkaMicroservice` (unlike the deployment on xref:aws-ingress.adoc[AWS] or xref:gcp-ingress.adoc[GCP] where the operator can fine tune all the necessary resources). The `Ingress` resources require cloud-specific tuning, you will need to refer to your kubernetes provider (e.g. minikube docs, Azure docs, etc...) in order to create an Ingress that can handle TLS-enabled, HTTP/2 traffic to serve gRPC externally.

Reviewer: Maybe something like this:
```suggestion
xref:gcp-install.adoc[Deploying with Google Cloud Platform (GCP)] and xref:aws-index.adoc[Deploying with Amazon Web Services (AWS)] integrate with cloud provider specific features, like databases, message brokers and ingress.

When you install the the Akka Cloud Platform operator on a custom Kubernetes cluster, it will not be in charge of creating the necessary `Ingress` instances to handle external traffic into your `AkkaMicroservice` (unlike the deployment on xref:aws-ingress.adoc[AWS] or xref:gcp-ingress.adoc[GCP] where the operator can fine tune all the necessary resources). The `Ingress` resources require cloud-specific tuning, you will need to refer to your kubernetes provider (e.g. minikube docs, Azure docs, etc...) in order to create an Ingress that can handle TLS-enabled, HTTP/2 traffic to serve gRPC externally.
```

= Custom Kubernetes Deployment

include::partial$include.adoc[]

Users with a
{lightbend-developer-docs}/introduction/getting-started/subscription-and-credentials.html[Lightbend Subscription {tab-icon}, window="tab"] have access to the Akka Cloud Platform operator to run on their own Kubernetes cluster.

> Make sure you read the xref:cloud-deployment.adoc[] and xref:deploy.adoc[] before proceeding to understand the basic deployment architecture and workflow.

Reviewer 4: ```suggestion
In order to best understand the basic deployment architecture and workflow, make sure you read the xref:cloud-deployment.adoc[] and xref:deploy.adoc[] before proceeding further.
```

== Prepare your cluster

> Make sure your `kubectl` context points to your local Kubernetes cluster (for example, {minikube}[minikube {tab-icon}, window="tab"]), or an external Kubernetes cluster (e.g. in https://docs.microsoft.com/en-ca/azure/aks/kubernetes-walkthrough[Azure {tab-icon}, window="tab"]).

Reviewer 4: ```suggestion
Verify that your `kubectl` context points to your local Kubernetes cluster (for example, {minikube}[minikube {tab-icon}, window="tab"]), or an external Kubernetes cluster (as illustrated in https://docs.microsoft.com/en-ca/azure/aks/kubernetes-walkthrough[Azure {tab-icon}, window="tab"]).
```
Too many 'make sure's'

Author: Thanks @thothmoses :-)

Next, make sure you have the akka-operator-helm chart:

[source,shell script]
----
helm repo add akka-operator-helm https://lightbend.github.io/akka-operator-helm/
helm repo update
----

Create environment variables with your {lightbend-com-account}/lightbend-platform/loginCredentials[Lightbend Login Credentials {tab-icon}, window="tab"]:

[source,shell script]
----
export LIGHTBEND_COMMERCIAL_USERNAME=12345678-aaa-bbb-cccc-ddddeeeeff0e@lightbend
export LIGHTBEND_COMMERCIAL_PASSWORD=abcdef1243568790ababab470556835429873569
----

Install the operator with the command:

[source,shell script,subs="attributes+"]
----
helm upgrade akka-operator akka-operator-helm/akka-operator \
  --set lightbendSubscription.username=$LIGHTBEND_COMMERCIAL_USERNAME \
  --set lightbendSubscription.password=$LIGHTBEND_COMMERCIAL_PASSWORD \
  --namespace lightbend \
  --version={akka-cloud-platform-version}  # <1>
----
<1> `{akka-cloud-platform-version}` is the latest available version.

The cluster is now ready for you to xref:deploy.adoc[deploy] `AkkaMicroservice` services.

== Known limitations

> Deploying the Akka Cloud Platform operator xref:gcp-install.adoc[with Google Cloud Platform (GCP)] or xref:aws-index.adoc[with Amazon Web Services (AWS)] enables integration with databases, message brokers and ingress specific to GCP or AWS.

Reviewer 4: ```suggestion
Deploying the Akka Cloud Platform operator xref:gcp-install.adoc[with Google Cloud Platform (GCP)] or xref:aws-index.adoc[with Amazon Web Services (AWS)] enables integration with databases, message brokers, and ingress specific to GCP or AWS.
```

> When you install the the Akka Cloud Platform operator on a custom Kubernetes cluster, it will not be in charge of creating the necessary `Ingress` instances to handle external traffic into your `AkkaMicroservice` (unlike the deployment on xref:aws-ingress.adoc[AWS] or xref:gcp-ingress.adoc[GCP] where the operator can fine tune all the necessary resources). The `Ingress` resources require cloud-specific tuning, you will need to refer to your kubernetes provider (e.g. minikube docs, Azure docs, etc...) in order to create an Ingress that can handle TLS-enabled, HTTP/2 traffic to serve gRPC externally.

Reviewer 4: ```suggestion
When you install the Akka Cloud Platform operator on a custom Kubernetes cluster, it will not be in charge of creating the `Ingress` instances required to handle external traffic into your `AkkaMicroservice`. This is unlike the deployment on xref:aws-ingress.adoc[AWS] or xref:gcp-ingress.adoc[GCP], where the operator can fine-tune all the necessary resources. The `Ingress` resources require cloud-specific tuning. You will need to refer to your Kubernetes provider (such as minikube docs, Azure docs) to create an Ingress that can handle TLS-enabled, HTTP/2 traffic to serve gRPC externally.
```
